### ISSUE TITLE
Support date/time on @colyseus/schema

### ISSUE THREAD

**Issue by dengzhaofun** [colyseus/colyseus#297]
Error: a 'string' was expected, but '"2020-01-28T15:33:25.708Z"' (Date) was provided in User#loginAt

  @type('string')
  @Column({comment: '最后登录时间'})
  loginAt: Date; // not work 

  @type('number')
  @Column({comment: '最后登录时间'})
  loginAt: Date; // not work

**Comment by endel** [comment]
Hi @dengzhaofun, `Date` is not supported as a type. You should use a timestamp instead. e.g. `Date.now()` or `date.getTime()`

**Comment by dengzhaofun** [comment]
Why not support the Date type, even if it helps to convert the number to display on the client. Forcing developers to change the data types of existing code is not a good solution

**Comment by endel** [comment]
Supporting `Date` in the future would be nice indeed, let's keep this issue open until we have it!

**Comment by vitalyrotari** [comment]
@endel `@type('string')` is casting to string?

I use `moment/luxon` for date & time

```TypeScript
import { DateTime, Duration } from "luxon";

DateTime.now().toString();  // 2021-03-05T19:26:46.628+02:00
Duration.fromMillis(30000).toString(); // PT30S
```
In my opinion `Date` is very primitive

```JavaScript
(new Date()).toString(); // Fri Mar 05 2021 19:31:51 GMT+0200 (Eastern European Standard Time)
```

**Comment by endel** [comment]
I'd prefer sending dates as number/unix timestamp rather than strings @vitalyrotari:

```typescript
class State extends Schema {
  @type("number") timestamp: number;
}

const state = new State();
state.timestamp = Date.now(); 
// => 1614970340065

const dateInstance = new Date(state.timestamp);
// => Fri Mar 05 2021 15:52:20 GMT-0300 (Brasilia Standard Time)

dateInstance.getTime() === state.timestamp
// => true
```

Eventually would be nice to convert seamlessly from unix timestamp to `Date` objects tho - the reason for not using strings from my perspective are that strings would consume slightly more data to transfer, and that unix timestamp is a format most client-side targets could easily convert into their own Date/DateTime object easily, for example:

- Lua: https://www.lua.org/pil/22.1.html
- C#: https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.fromunixtimeseconds?redirectedfrom=MSDN&view=net-5.0#System_DateTimeOffset_FromUnixTimeSeconds_System_Int64_

**Comment by vitalyrotari** [comment]
@endel unix timestamp doesn't preserve time zone, this is why I use `ISO 8601`. So if timezone is not critical, yes we can use unix timestamp :)

**Comment by endel** [comment]
@vitalyrotari I think it does! check out this answer on Stack Overflow: https://stackoverflow.com/questions/23062515/do-unix-timestamps-change-across-timezones/23062640#23062640

> "The definition of UNIX timestamp is timezone independent. The timestamp is the number of seconds (or milliseconds) elapsed since an absolute point in time, midnight of Jan 1 1970 in UTC time. (UTC is Greenwich Mean Time without Daylight Savings time adjustments.) Regardless of your timezone, a timestamp represents a moment that is the same everywhere. Of course you can convert back and forth to a local timezone representation (time 1397484936 is such-and-such local time in New York, or some other local time in Djakarta) if you want"

**Comment by endel** [comment]
Oh I understand what you mean now, it is the same time but it does not specify which timezone that time is, that's right!

**Comment by vitalyrotari** [comment]
@endel yep :)  So the bulletproof option is `ISO 8601`, but that's if accuracy is required.

**Comment by hunkydoryrepair** [comment]
But, `Date` object is javascript doesn't preserve timezone, either.
So if you want ISO with timezone, you need to use a string. Date object doesn't help you. Converting Date => string => Date will not provide timezone information as you can only get UTC or local time from a Date object.

**Comment by endel** [comment]
We have rudimentary support for custom types since version 0.16: https://docs.colyseus.io/state/advanced-usage#custom-types-and-encoding

I'm going to close this issue but a pull request on [colyseus/schema](https://github.com/colyseus/schema) is welcome if anybody wants to give it a shot. We'd need to have the decoder in all SDKs, though. Cheers!
